Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp

```diff
@@ -239,6 +252,49 @@ void ImageBufferIOSurfaceBackend::ensureNativeImagesHaveCopiedBackingStore()
     flushContext();
 }
 
+void* ImageBufferIOSurfaceBackend::updateBitmapCacheForRead()
+{
+    if (!m_bitmapCacheLock.has_value()) {
+        flushContext();
+        m_bitmapCacheLock = IOSurface::Locker { *m_surface, IOSurface::Locker::AccessMode::ReadOnly };
+    }
+    return m_bitmapCacheLock->surfaceBaseAddress();
+}
+
+CGImageRef ImageBufferIOSurfaceBackend::updateBitmapCacheImage()
+{
+    if (m_bitmapCacheImage)
+        return m_bitmapCacheImage.get();
+    void* data = updateBitmapCacheForRead();
+    if (!data)
+        return nullptr;
+    auto configuration = m_surface->bitmapConfiguration();
+    auto size = m_surface->size();
+    // FIXME: we can cache also the context and just use set data operation, in later commits.
+    auto context = adoptCF(CGBitmapContextCreate(data, size.width(), size.height(), configuration.bitsPerComponent, m_surface->bytesPerRow(), m_surface->colorSpace().platformColorSpace(), configuration.bitmapInfo));
+    if (!context)
+        return nullptr;
+    m_bitmapCacheImage = adoptCF(CGBitmapContextCreateImage(context.get()));
```

Comment:
How much is this faster than calling `IOSurface::createImage()`? And it is considerably faster why do not we move this code in the `IOSurface` itself?

Reply:
This is (roughly) the same. The difference is that here we get access to the actual bitmap image, to which we can apply future options. We cannot apply those options to the IOSurface -backed image and expect them to get propagated to the eventual bitmap copy that the IOSurface -backed image needs to make. Or we could, but it doesn't work that way currently and we'd then need to go and convince this is the case and then fix the platform  and then merge the fixes to target platforms and ..